rir/src/compiler/analysis/refrence_count.h

```diff
@@ -0,0 +1,109 @@
+#ifndef PIR_REFERENCE_COUNT_H
+#define PIR_REFERENCE_COUNT_H
+
+#include "../pir/pir_impl.h"
+#include "generic_static_analysis.h"
+
+namespace rir {
+namespace pir {
+
+struct AUses {
+    std::unordered_map<Instruction*, size_t> uses;
+    AbstractResult mergeExit(const AUses& other) { return merge(other); }
+    AbstractResult merge(const AUses& other) {
+        AbstractResult res;
+        for (auto u : other.uses) {
+            if (!uses.count(u.first)) {
+                uses.emplace(u);
+                res.update();
+            } else if (uses.at(u.first) < u.second) {
+                // used in both branches, use count is the maximum
+                uses.at(u.first) = u.second;
+                res.update();
+            }
+        }
+        return res;
+    }
+    void print(std::ostream&, bool) const {
+        // TODO
+    }
+};
+
+class StaticReferenceCount : public StaticAnalysis<AUses> {
+  public:
+    StaticReferenceCount(ClosureVersion* cls, LogStream& log)
+        : StaticAnalysis("StaticReferenceCountAnalysis", cls, cls, log) {
+        bool changed = true;
+        while (changed) {
+            changed = false;
+
+            Visitor::run(code->entry, [&](Instruction* i) {
+                // Recursively enumerate all actual values a phi might contain
+                if (Phi::Cast(i)) {
+                    i->eachArg([&](Value* v) {
+                        if (auto a = Instruction::Cast(v)) {
+                            if (Phi::Cast(a)) {
+                                for (auto otherAlias : alias[a]) {
+                                    if (!alias[i].includes(otherAlias)) {
+                                        changed = true;
+                                        alias[i].insert(otherAlias);
+                                    }
+                                }
+                            } else {
+                                if (!alias[i].includes(a)) {
+                                    changed = true;
+                                    alias[i].insert(a);
+                                }
+                            }
+                        }
+                    });
+                }
+            });
+        }
+    }
+
+  protected:
+    std::unordered_map<Instruction*, SmallSet<Instruction*>> alias;
+
+    AbstractResult apply(AUses& state, Instruction* i) const override {
+        AbstractResult res;
+        if (Phi::Cast(i))
+            return res;
+
+        if (i->needsReferenceCount()) {
+            // This value was used only once in a loop, we can thus
+            // reset the count on redefinition.
+            if (state.uses.count(i) && state.uses.at(i) == 1) {
+                state.uses[i] = 0;
+                res.update();
+            }
+        }
+
+        auto count = [&](Instruction* i) {
+            auto& use = state.uses[i];
+            if (use < 2) {
```

Comment:
BTW, `NAMED_MAX` is 3 in our version of R, but that doesn't matter here I presume?

Reply:
hmm, good point. at least rir only discriminates between 0,1 and more. I don't know of any situation where we could do something else if it is 2 vs. 3, do you?

Reply:
Not really.. the only thing is we could sometimes decrement the named maybe? Because 2 really should mean 2 names of the value, not at least 2.

Reply:
yeah, this might be an interesting optimization to add, if we find such cases